### PR TITLE
Use robust boundary state for Geant4 navigator wrapper

### PR DESCRIPTION
**Pull request by sethrj**
This adds a `GeoStatus` to the Geant4 geometry wrapper as was done in #2312 . The upshot is that the G4 wrapper can handle the single in-and-out transaction that's needed for optical surfaces and field propagation (and it can handle any number of direction changes while in the same volume).

While testing and comparing I also found an edge case that could cause trouble if we're not careful: changing direction to be perpendicular to a surface normal. Doing this correctly would require surface concavity in addition to the normal direction, and would have no good solution for planar surfaces (or other edge cases e.g. cones) so I think we should error if that happens, which we now do.

I updated and rearranged a couple of the reentrant tests so that failures are more easily handled.

### CODE REVIEW

**Comment by sethrj** [comment]
@elliottbiondo This one also is ready for review, though I know you've not got too much experience with the Geant4 navigator itself.

@mrguilima I don't know how much you've worked with the G4Navigator but this could be interesting for you too.

**Comment by sethrj** [comment]
@elliottbiondo could you also please review this week?

**Comment by elliottbiondo** [comment]
I will try! Sorry, extra busy this week.

File: src/geocel/g4/GeantGeoTrackView.hh [hunk 1]
```diff
@@ -555,6 +627,34 @@ void GeantGeoTrackView::move_internal(Real3 const& pos)
...
+        auto new_dot = dot_product(norm, newdir);
+        if (CELER_UNLIKELY(new_dot == 0))
+        {
+            CELER_LOG_LOCAL(error)
+                << "track direction cannot change to " << newdir
+                << " which is perpendicular to the current surface normal";
+            this->geo_status(GeoStatus::error);
+            return;
+        }
+        else if ((dot_product(norm, newdir) > 0)
```
Comment (mrguilima) [inline src/geocel/g4/GeantGeoTrackView.hh]:
```suggestion
        else if ((new_dot > 0)
```

**Review by mrguilima — APPROVED** [review]
LGTM, but I haven't looked a lot at the logic of the tests in detail yet.
Thanks!

**Review by elliottbiondo — APPROVED** [review]
